node/service/src/lib.rs

```diff
@@ -782,6 +783,23 @@ where
 					Ok((time, parachain_inherent, author, randomness))
 				}
 			};
+			// must be defined here to access client runtime API in this function
+			// type Digests = Option<DigestItem> should work
+			let additional_digests_provider =  |nimbus_id: nimbus_primitives::NimbusId, parent: Hash| -> Option<sp_runtime::generic::DigestItem> {
+				let relay_slot_number: polkadot_primitives::v2::Slot = todo!(); // get using runtime API
+				let relay_storage_root: Hash = todo!(); // get using runtime API
+				let key: session_keys_primitives::VrfId = todo!(); // get using runtime API VrfKeyLookup
+				if let Some(vrf_pre_digest) = moonbeam_vrf::vrf_pre_digest::<Hash>(
+					relay_slot_number,
+					relay_storage_root,
+					key,
+					&*&keystore,
+				) {
+					Some(CompatibleDigestItem::vrf_pre_digest(vrf_pre_digest))
+				} else {
+					None
+				}
+			};
```

Comment:
@librelois I intended to implement `DigestsProvider` with a closure like this, but the keystore cannot be borrowed in this way:
```rust
(amar-per-block-vrf)⚡ [101] % bb                                                                                             ~/ong/ps/mb/node/service
   Compiling moonbeam-service v0.24.0 (/Users/4meta5/ong/ps/mb/node/service)
warning: unreachable statement
   --> node/service/src/lib.rs:790:5
    |
789 |                 let relay_slot_number: polkadot_primitives::v2::Slot = todo!(); // get using runtime API
    |                                                                        ------- any code following this expression is unreachable
790 |                 let relay_storage_root: Hash = todo!(); // get using runtime API
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unreachable statement
    |
    = note: `#[warn(unreachable_code)]` on by default

error[E0373]: closure may outlive the current function, but it borrows `keystore`, which is owned by the current function
   --> node/service/src/lib.rs:788:39
    |
788 |             let additional_digests_provider =  |nimbus_id: nimbus_primitives::NimbusId, parent: Hash| -> Option<sp_runtime::generic::D...
    |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ may outlive borrowed value `keystore`
...
792 |                 let current_keystore = keystore.clone();
    |                                        -------- `keystore` is borrowed here
    |
note: function requires argument type to outlive `'static`
   --> node/service/src/lib.rs:805:7
    |
805 |               Ok(NimbusConsensus::build(BuildNimbusConsensusParams {
    |  ________________^
806 | |                 para_id: id,
807 | |                 proposer_factory,
808 | |                 block_import: client.clone(),
...   |
813 | |                 additional_digests_provider,
814 | |             }))
    | |______________^
help: to force the closure to take ownership of `keystore` (and any other referenced variables), use the `move` keyword
    |
788 |             let additional_digests_provider =  move |nimbus_id: nimbus_primitives::NimbusId, parent: Hash| -> Option<sp_runtime::generic::DigestItem> {
    |                                                ++++

error[E0505]: cannot move out of `keystore` because it is borrowed
   --> node/service/src/lib.rs:810:5
    |
788 |               let additional_digests_provider =  |nimbus_id: nimbus_primitives::NimbusId, parent: Hash| -> Option<sp_runtime::generic::D...
    |                                                  ------------------------------------------------------------------------------------------------- borrow of `keystore` occurs here
...
792 |                   let current_keystore = keystore.clone();
    |                                          -------- borrow occurs due to use in closure
...
805 |               Ok(NimbusConsensus::build(BuildNimbusConsensusParams {
    |  ________________-
806 | |                 para_id: id,
807 | |                 proposer_factory,
808 | |                 block_import: client.clone(),
809 | |                 parachain_client: client.clone(),
810 | |                 keystore,
    | |                 ^^^^^^^^ move out of `keystore` occurs here
...   |
813 | |                 additional_digests_provider,
814 | |             }))
    | |______________- argument requires that `keystore` is borrowed for `'static`

Some errors have detailed explanations: E0373, E0505.
For more information about an error, try `rustc --explain E0373`.
warning: `moonbeam-service` (lib) generated 1 warning
error: could not compile `moonbeam-service` due to 2 previous errors; 1 warning emitted
```

It feels like it would be easier if we added a parameter to the `provide_digests` function in nimbus which would include the current keystore and client (to get the Runtime API values).

Reply:
@4meta5 you just need to clone the keystore before the closure AND use the move keyword. Same for the client.